Installation issue: hdf5-vol-async

### Steps to reproduce the issue

```console
$ spack spec -I hdf5-vol-async
Input spec
--------------------------------
 -   hdf5-vol-async

Concretized
--------------------------------
 -   hdf5-vol-async@1.3%apple-clang@14.0.0~ipo build_system=cmake build_type=RelWithDebInfo arch=darwin-monterey-m1
[+]      ^argobots@main%apple-clang@14.0.0~affinity~debug+perf~stackunwind~tool~valgrind build_system=autotools stackguard=none arch=darwin-monterey-m1
[+]          ^autoconf@2.69%apple-clang@14.0.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=darwin-monterey-m1
[+]              ^gnuconfig@2021-08-14%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
[+]          ^automake@1.16.5%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]          ^gnuconfig@2022-09-17%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
[+]          ^libtool@2.4.7%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]          ^m4@1.4.19%apple-clang@14.0.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=darwin-monterey-m1
[+]              ^diffutils@3.8%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]                  ^libiconv@1.16%apple-clang@14.0.0 build_system=autotools libs=shared,static arch=darwin-monterey-m1
[+]              ^libsigsegv@2.13%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]      ^cmake@3.25.0%apple-clang@14.0.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=darwin-monterey-m1
[+]          ^ncurses@6.3%apple-clang@14.0.0~symlinks+termlib abi=none build_system=autotools arch=darwin-monterey-m1
[+]          ^openssl@1.1.1s%apple-clang@14.0.0~docs~shared build_system=generic certs=mozilla arch=darwin-monterey-m1
[+]              ^ca-certificates-mozilla@2022-10-11%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
[+]      ^hdf5@1.13.3%apple-clang@14.0.0~cxx~fortran~hl~ipo~java+mpi+shared~szip+threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo arch=darwin-monterey-m1
[+]          ^pkgconf@1.8.0%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]          ^zlib@1.2.13%apple-clang@14.0.0+optimize+pic+shared build_system=makefile arch=darwin-monterey-m1
[+]      ^openmpi@4.1.4%apple-clang@14.0.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker+romio+rsh~singularity+static+vt+wrapper-rpath build_system=autotools fabrics=none schedulers=none arch=darwin-monterey-m1
[+]          ^hwloc@2.8.0%apple-clang@14.0.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl~pci~rocm build_system=autotools libs=shared,static arch=darwin-monterey-m1
[+]              ^libxml2@2.10.1%apple-clang@14.0.0~python build_system=autotools arch=darwin-monterey-m1
[+]                  ^xz@5.2.7%apple-clang@14.0.0~pic build_system=autotools libs=shared,static arch=darwin-monterey-m1
[+]          ^openssh@9.1p1%apple-clang@14.0.0+gssapi build_system=autotools arch=darwin-monterey-m1
[+]              ^krb5@1.20.1%apple-clang@14.0.0+shared build_system=autotools arch=darwin-monterey-m1
[+]                  ^bison@3.8.2%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]                  ^gettext@0.21.1%apple-clang@14.0.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=darwin-monterey-m1
[+]                      ^tar@1.34%apple-clang@14.0.0 build_system=autotools zip=pigz arch=darwin-monterey-m1
[+]                          ^pigz@2.7%apple-clang@14.0.0 build_system=makefile arch=darwin-monterey-m1
[+]                          ^zstd@1.5.2%apple-clang@14.0.0+programs build_system=makefile compression=none libs=shared,static arch=darwin-monterey-m1
[+]              ^libedit@3.1-20210216%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]              ^libxcrypt@4.4.31%apple-clang@14.0.0~obsolete_api build_system=autotools arch=darwin-monterey-m1
[+]          ^perl@5.36.0%apple-clang@14.0.0+cpanm+shared+threads build_system=generic arch=darwin-monterey-m1
[+]              ^berkeley-db@18.1.40%apple-clang@14.0.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=darwin-monterey-m1
[+]              ^bzip2@1.0.8%apple-clang@14.0.0~debug~pic+shared build_system=generic arch=darwin-monterey-m1
[+]              ^gdbm@1.23%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]                  ^readline@8.1.2%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
[+]          ^pmix@4.1.2%apple-clang@14.0.0~docs+pmi_backwards_compatibility~python~restful build_system=autotools arch=darwin-monterey-m1
[+]              ^libevent@2.1.12%apple-clang@14.0.0+openssl build_system=autotools arch=darwin-monterey-m1
```


### Error message

```
/var/folders/j1/68dlgpr91vlgs26vty2c8xk80000gn/T/ajstewart/spack-stage/spack-stage-hdf5-vol-async-1.3-t6ezorzrly62th5zpjnzmhcd66yp2i33/spack-src/src/h5_async_vol.c:8614:87: error: too few arguments to function call, expected 9, have 8
                                  args->file_space_id, args->plist_id, args->buf, NULL);
                                                                                      ^
```

### Information on your system

* **Spack:** 0.20.0.dev0 (fe995542ab615ad16e13e1143ea3d5a5b5a0eb94)
* **Python:** 3.10.8
* **Platform:** darwin-monterey-m1
* **Concretizer:** clingo


### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/10155768/spack-build-out.txt)
* [spack-build-env-mods.txt](https://github.com/spack/spack/files/10155769/spack-build-env-mods.txt)

@houjun @hyoklee @jeanbez 
@becker33 @bvanessen @alalazo 

I first noticed this build issue in #34299. The E4S pipeline is broken on develop because of this bug, so any PR that modifies one of the packages that hdf5-vol-async depends on will fail CI. See https://gitlab.spack.io/spack/spack/-/jobs/4817633 for the first occurrence of this bug. Not sure how this relates to #33939, but I'm guessing that PR in combination with another PR caused the issue.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
+++ b/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
@@ -24,8 +24,9 @@ class Hdf5VolAsync(CMakePackage):
 
     depends_on("mpi")
     depends_on("argobots@main")
-    depends_on("hdf5@1.13: +mpi +threadsafe")
-
+    depends_on("hdf5@1.13.2 +mpi +threadsafe")
+    depends_on("hdf5@1.13.3: +mpi +threadsafe", when="@1.4:")
+    
     def setup_run_environment(self, env):
         env.set("HDF5_PLUGIN_PATH", self.spec.prefix.lib)
         vol_connector = "async under_vol=0;under_info=[]"

--- a/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
+++ b/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
@@ -24,8 +24,9 @@ class Hdf5VolAsync(CMakePackage):
 
     depends_on("mpi")
     depends_on("argobots@main")
-    depends_on("hdf5@1.13.0:1.13.2 +mpi +threadsafe")
-    depends_on("hdf5@1.13.3: +mpi +threadsafe", when="@1.4:")
+    depends_on("hdf5 +mpi +threadsafe")
+    depends_on("hdf5@1.13.0:1.13.2", when="@:1.3")
+    depends_on("hdf5@1.13.3:", when="@1.4:")
 
     def setup_run_environment(self, env):
         env.set("HDF5_PLUGIN_PATH", self.spec.prefix.lib)

--- a/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
+++ b/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
@@ -26,7 +26,7 @@ class Hdf5VolAsync(CMakePackage):
     depends_on("argobots@main")
     depends_on("hdf5@1.13.2 +mpi +threadsafe")
     depends_on("hdf5@1.13.3: +mpi +threadsafe", when="@1.4:")
-    
+
     def setup_run_environment(self, env):
         env.set("HDF5_PLUGIN_PATH", self.spec.prefix.lib)
         vol_connector = "async under_vol=0;under_info=[]"

--- a/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
+++ b/var/spack/repos/builtin/packages/hdf5-vol-async/package.py
@@ -24,7 +24,7 @@ class Hdf5VolAsync(CMakePackage):
 
     depends_on("mpi")
     depends_on("argobots@main")
-    depends_on("hdf5@1.13.2 +mpi +threadsafe")
+    depends_on("hdf5@1.13.0:1.13.2 +mpi +threadsafe")
     depends_on("hdf5@1.13.3: +mpi +threadsafe", when="@1.4:")
 
     def setup_run_environment(self, env):

